### PR TITLE
feat!: add download utils function for shared aggregated and algo state

### PR DESCRIPTION
**Pull request by ThibaultFy**
## Related issue

fixes FL-633

## Companion PR

- https://github.com/Substra/substra-documentation/pull/313

## Summary

To ease analytics usage for users, we implement tool function such as:
`download_algo_state`
`download_shared_state`
`download_aggregated_state`
It is now possible to download both from round or rank.

See https://www.notion.so/owkin-fdn/Federated-Analytics-in-SubstraFL-5685c4f2549748a89a3ad2e314cde886?pvs=4 for more info.

## Please check if the PR fulfills these requirements

- [x] If the feature has an impact on the user experience, the [changelog](https://github.com/substra/substrafl/blob/main/CHANGELOG.md) has been updated
- [x] Tests for the changes have been added (for bug fixes / features)
- [x] Docs have been added / updated (for bug fixes / features)
- [x] The commit message follows the [conventional commit](https://www.conventionalcommits.org/en/v1.0.0/) specification

### CODE REVIEW

**Comment by Owlfred** [comment]
End to end tests: :hourglass_flowing_sand: **STARTING**

File: docs/api/model_loading.rst [hunk 1]
```diff
@@ -1,9 +1,6 @@
 Model Loading
```
Comment (SdgJlbl) [inline docs/api/model_loading.rst:1]:
```suggestion
Model Downloading
```
If we are going to nitpick 🙃

File: substrafl/model_loading.py [hunk 2]
```diff
@@ -147,93 +144,191 @@ def _get_train_task_from_round(client: substra.Client, compute_plan_key: str, ro
...
+        local_tasks = client.list_task(filters=filters, order_by="end_date")
+        local_tagged_tasks = [t for t in local_tasks if t.tag == tag]
+
+    if len(local_tagged_tasks) == 0:
+        raise exceptions.TaskNotFoundError(
+            f"The given compute plan `{compute_plan_key}` has no {tag} task of rank {rank_idx} "
+            f"hosted on the organization {org_id}"
+        )
+
+    elif len(local_tagged_tasks) > 1:
```
Comment (SdgJlbl) [inline substrafl/model_loading.py:201]: If `rank_idx` is None, shouldn't we return the last task, in case of multiples, rather than raising? 
At least, that would be the behaviour I expect reading the docstring 🤔

File: substrafl/model_loading.py [hunk 3]
```diff
@@ -147,93 +144,191 @@ def _get_train_task_from_round(client: substra.Client, compute_plan_key: str, ro
...
+    if not remote:
+        instance = remote_struct.get_instance()
+    else:
+        instance = remote_struct.get_remote_instance()
+
+    return instance
+
 
-    return my_algo
+def _load_from_files(input_folder: os.PathLike, remote: bool = False) -> Any:
```
Comment (SdgJlbl) [inline substrafl/model_loading.py:239]: why not use `pathlib.Path`? I have the impression it's the standard in substrafl

File: substrafl/model_loading.py [hunk 4]
```diff
@@ -147,93 +144,191 @@ def _get_train_task_from_round(client: substra.Client, compute_plan_key: str, ro
...
+
+    try:
+        instance = _load_instance(gz_path=folder / metadata[FUNCTION_DICT_KEY], extraction_folder=folder, remote=remote)
+
+        if not remote:
+            loaded_instance = instance.load_local_state(folder / metadata[MODEL_DICT_KEY])
+        else:
+            loaded_instance = instance.load_model(folder / metadata[MODEL_DICT_KEY])
+
+    except ModuleNotFoundError as e:
```
Comment (SdgJlbl) [inline substrafl/model_loading.py:274]: do we except the first or the second half to raise? would it be possible to have a narrower try / except encapsulation?

File: substrafl/model_loading.py [hunk 5]
```diff
@@ -147,93 +144,191 @@ def _get_train_task_from_round(client: substra.Client, compute_plan_key: str, ro
...
         - a metadata.json
 
-    Important:
-        This function supports only strategies with one train task for a given organization and round.
-
     Args:
         client (substra.Client): Substra client where to fetch the model from.
         compute_plan_key (str): Compute plan key to fetch the model from.
         dest_folder (os.PathLike): Folder where to download the files.
+        task_type (str): Type of the task to fetch. Allowed value are "init", "train" and "aggregate".
```
Comment (SdgJlbl) [inline substrafl/model_loading.py:313]: minor but maybe we want that to be an `Enum`? 🤔

File: substrafl/model_loading.py [hunk 6]
```diff
@@ -242,71 +337,138 @@ def download_algo_files(
...
-    # Get the train task associated to user inputs
-    train_task = _get_train_task_from_round(client=client, compute_plan_key=compute_plan_key, round_idx=round_idx)
-
-    if train_task.status is not Status.done:
-        raise UnfinishedTrainTaskError(
-            f"Can't download algo files form task {train_task.key} as it is " f"in status {train_task.status}"
+    # Get the task associated to user inputs
+    if round_idx is not None and rank_idx is not None:
+        raise exceptions.ArgumentConflictError("Only one out of round_idx and rank_idx must be specified.")
+    elif round_idx is not None:
```
Comment (SdgJlbl) [inline substrafl/model_loading.py:343]: Minor but I would restart with a regular `if` here, given that we have raised in the other case.

File: substrafl/model_loading.py [hunk 7]
```diff
@@ -242,71 +337,138 @@ def download_algo_files(
...
-    try:
-        algo = _load_algo(algo_path=folder / metadata[ALGO_DICT_KEY], extraction_folder=folder)
-        local_state = algo.load_local_state(folder / metadata[LOCAL_STATE_DICT_KEY])
+    Args:
+        client (substra.Client): Substra client where to fetch the model from.
+        compute_plan_key (str): Compute plan key to fetch the model from.
+        round_idx (Optional[int], None): Round of the strategy to fetch the model from. If set to ``None``,
+            the rank_idx will be used. (Defaults to None)
+        rank_idx (Optional[int], None): Rank of the strategy to fetch the model from. If set to ``None``, the last task
+            (with the highest rank) will be used. (Defaults to None)
```
Comment (SdgJlbl) [inline substrafl/model_loading.py:459]: missing return documentation

File: tests/algorithms/pytorch/test_newton_raphson.py [hunk 8]
```diff
@@ -355,11 +354,13 @@ def test_pytorch_nr_algo_performance(
...
 @pytest.mark.slow
 @pytest.mark.substra
-def test_download_load_algo(network, compute_plan, session_dir, nr_test_data, mae, rtol):
-    download_algo_files(
-        client=network.clients[0], compute_plan_key=compute_plan.key, round_idx=None, dest_folder=session_dir
+def test_download_load_algo(network, compute_plan, nr_test_data, mae, rtol):
+    algo = download_algo_state(
+        client=network.clients[0],
+        compute_plan_key=compute_plan.key,
+        round_idx=NUM_ROUNDS,
```
Comment (SdgJlbl) [inline tests/algorithms/pytorch/test_newton_raphson.py:361]: I might be missing something, but wouldn't the equivalent of the previous test be with `round_idx=None`?

File: tests/algorithms/pytorch/test_scaffold.py [hunk 9]
```diff
@@ -346,11 +344,13 @@ def _local_train(self, train_dataset):
...
 @pytest.mark.slow
 @pytest.mark.substra
-def test_download_load_algo(network, compute_plan, session_dir, test_linear_data_samples, mae, rtol):
-    download_algo_files(
-        client=network.clients[0], compute_plan_key=compute_plan.key, round_idx=None, dest_folder=session_dir
+def test_download_load_algo(network, compute_plan, test_linear_data_samples, mae, rtol):
+    algo = download_algo_state(
+        client=network.clients[0],
+        compute_plan_key=compute_plan.key,
+        round_idx=NUM_ROUNDS,
```
Comment (SdgJlbl) [inline tests/algorithms/pytorch/test_scaffold.py:351]: same as for NR

File: tests/test_model_loading.py [hunk 10]
```diff
@@ -164,73 +196,144 @@ def load_local_state(self, path):
...
+@pytest.mark.parametrize(
+    "round_idx, rank_idx, expectation",
+    (
+        [None, None, does_not_raise()],
+        [1, 1, pytest.raises(exceptions.ArgumentConflictError)],
+        [1, None, does_not_raise()],
+        [None, 1, does_not_raise()],
+    ),
+)
+def test_round_idx_vs_rank_idx(
```
Comment (SdgJlbl) [inline tests/test_model_loading.py:234]: can you check that actually the right subfunction is being called as well?

File: substrafl/model_loading.py (hunk 2 above)
Comment (ThibaultFy) [reply substrafl/model_loading.py:201]: Nice catch thanks a lot !

File: substrafl/model_loading.py (hunk 5 above)
Comment (ThibaultFy) [reply substrafl/model_loading.py]: 100% agree, thanks. I created a schemas.py at the root of substrafl to enum the `TaskType`. It seems to me the best place to put it, but not 100% sure

File: tests/algorithms/pytorch/test_newton_raphson.py (hunk 8 above)
Comment (ThibaultFy) [reply tests/algorithms/pytorch/test_newton_raphson.py:361]: your not, I used it to test sorry :) I changed it back

File: substrafl/model_loading.py (hunk 5 above)
Comment (SdgJlbl) [reply substrafl/model_loading.py]: It seems reasonable to me too

**Review by SdgJlbl — APPROVED** [review]
Still unconvinced about the need for a custom `ArgumentConflictError` exception, otherwise LGTM 👍
